Topics/User_Experience/Fuzzy_Search.md

```diff
@@ -0,0 +1,52 @@
+ 
+## Implementation: Levenshtein distance
+In this section, we discuss the [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance) and how to use it in implementing fuzzy search.
+
+Levenshtein is a popular distance metric to compare the similarity between a query string and a target string. Intuitively, it counts the fewest number of times a character in the query must be deleted, inserted, or replaced to reach the target. For instance, if the query is `Brian`, its Levenshtein distance to `Bryan` is 1 as it only needs to replace `i` with `y` to reach the target. Notice that this can also be done in two operations by deleting `i` and inserting `y`, but this is suboptimal. Targets with a lower Levenshtein score are considered closer to the query.
+
+The algorithm for computing Levenshtein distance is not overly complicated and can be a fun exercise in designing recursive functions. [This GeeksforGeeks article](https://www.geeksforgeeks.org/introduction-to-levenshtein-distance/) describes several possible implementations in different programming languages should you decide to code it yourself.
+
+That being said, you can also find out-of-the-box solutions online that are likely better optimized. For applications built in JavaScript, the [js-levenshtein](https://www.npmjs.com/package/js-levenshtein) library is quite efficient and easy to use:
+- Install it by running `npm install --save js-levenshtein`.
+- In your code, import the module using `const levenshtein = require('js-levenshtein');`
+- Compute Levenshtein distance using `levenshtein('queryString', 'targetString')`
+
+The following code snippet sorts a list of names by their Levenshtein proximity to `Brian`:
+```
+const query = 'Brian';  // Define the query string
+const names = ['Bryan', 'Brian', 'Bob', 'Bryant', 'HarrisonFord'];  // Define the list of strings to search from
+names.sort((a, b) => levenshtein(query, a) - levenshtein(query, b));  // Sort the list based on Levenshtein distance
+
+// Result: ['Brian', 'Bryan', 'Bryant', 'Bob', 'HarrisonFord'] with distances [0, 1, 2, 4, 9] respectively
+```
+
+The above strategy can be used, as an example, to take in an input query string from a search bar component and sort the list of results to display before they are rendered back to the user. Additionally, instead of just sorting by distance, you can devise a threshold so that only results below a certain distance are included, to prevent extremely distant results like `HarrisonFord` from appearing at all, and those that do appear are sorted as before.
```

Comment:
```suggestion
The above strategy can be used, as an example, to take in an input query string from a search bar component and sort the list of results to display before they are rendered back to the user. Additionally, instead of just sorting by distance, you can devise a threshold so that only results below a certain distance are included, to prevent extremely distant results like `HarrisonFord` from appearing at all.
```
You could probably remove the last phrase. (nitpick)

Reply:
Sure, will do. Actually the interesting thing is that as I was writing that I wasn't sure if it was extraneous, so this is a good confirmation that it should be removed. Thanks!